data_processors/pipeline/lambdas/orchestrator.py

```diff
@@ -345,63 +344,47 @@ def prepare_germline_jobs(this_batch: Batch, this_batch_run: BatchRun, this_sqr:
-        if sample_library_name.startswith("NTC_"):
-            logger.warning(f"SKIP NTC SAMPLE '{sample_library_name}' GERMLINE WORKFLOW LAUNCH.")
+        # Get the metadata for the library
+        # NOTE: this will use the library base ID (i.e. without topup/rerun extension), as the metadata is the same
+        rglb = row['rglb']
+        print(f"DEBUG {rglb}")
+        lib_metadata: LabMetadata = LabMetadata.objects.get(library_id=rglb)
+        # make sure we have recognised sample (e.g. not undetermined)
+        if not lib_metadata:
+            logger.error(f"SKIP GERMLINE workflow for {rgid}. No metadata for {rglb}, this should not happen!")
             continue
 
-        # Get sample metadata
-        library_metadata: pd.DataFrame = metadata_df.query(f"sample=='{sample_library_name}'")
-
-        # Check library metadata is not empty
-        if library_metadata.empty:
-            logger.warning(f"Could not get library metadata for sample '{sample_library_name}'. "
-                           f"Skipping sample.")
+        # skip negative control samples
+        if lib_metadata.phenotype.lower() == LabMetadataPhenotype.N_CONTROL.value.lower():
+            logger.info(f"SKIP GERMLINE workflow for '{rgid}'. Negative-control.")
             continue
 
         # Skip samples where metadata workflow is set to manual
-        if library_metadata["workflow"].unique().item() == "manual":
+        if lib_metadata.workflow.lower() == LabMetadataWorkflow.MANUAL.value.lower():
             # We do not pursue manual samples
-            logger.info(f"Skipping sample '{sample_library_name}'. "
-                        f"Workflow column is set to manual for at least one library name")
+            logger.info(f"SKIP GERMLINE workflow for '{rgid}'. Workflow set to manual.")
             continue
 
-        # iterate through libraries for this sample and collect their assay types
-        assay_type = library_metadata["type"].unique().item()
-
         # skip germline if assay type is not WGS
-        if assay_type != "WGS":
-            logger.warning(f"SKIP {assay_type} SAMPLE '{sample_library_name}' GERMLINE WORKFLOW LAUNCH.")
+        if lib_metadata.type.lower() != LabMetadataType.WGS.value.lower():
+            logger.warning(f"SKIP GERMLINE workflow for '{rgid}'. 'WGS' != '{lib_metadata.type}'.")
             continue
 
         # convert read_1 and read_2 to cwl file location dict format
-        sample_df["read_1"] = sample_df["read_1"].apply(cwl_file_path_as_string_to_dict)
-        sample_df["read_2"] = sample_df["read_2"].apply(cwl_file_path_as_string_to_dict)
+        row["read_1"] = cwl_file_path_as_string_to_dict(row["read_1"])
+        row["read_2"] = cwl_file_path_as_string_to_dict(row["read_2"])
 
         job = {
-            "sample_name": sample_library_name,
-            "fastq_list_rows": sample_df.drop(columns=["sample_library_names"]).to_dict(orient="records"),
+            "sample_name": rgid,
```

Comment:
Ah, `rgid` is in a form of `index1.index2.lane.seq_name.uniq_id` , pls [see here](https://github.com/umccr/data-portal-apis/blob/dev/data_processors/pipeline/lambdas/fastq_list_row.py#L126)

Or, check it in prod db for example as follows:
```
select * from data_portal_fastqlistrow order by id desc limit 10;
```

Reply:
Yes, again probably one for discussion.

Here I just wanted a unique enough ID so we can link the workflow with the generated FASTQs (a misuse of that attribute, I admit). We've discussed that issue before in the DevOps meeting, how to link workflows to libraries/library runs. Perhaps we can extend the workflow model?
Again, happy to discuss.

Reply:
I think `rgid` is too unique here. `rgid.rsplit(".", 1)[-1]` is probably the most appropriate unique ID for a sample name.

Reply:
`rgid` will be different for each lane. A sample may exist over multiple lanes.

Reply:
Hm, true. 
However, a sample over multiple lanes would create multiple FASTQs, right? And would that then not fall into the same category we discussed earlier of not merging topups even if on the same run? How would a sample split over multiple lanes be different from a topup on the same run?
Sorry, this is still screwing with my head!

Reply:
LoL; I'd vote; let us push this through and tinker. We always have DEV environment and can get tested there. Rinse & spin; keep the wheel spinning, I'd say!

Reply:
> How would a sample split over multiple lanes be different from a topup on the same run?

Multiple fastqs on the same run will have the same rgsm value if they're from the same sample but split over multiple lanes. Topups and non top ups would have different rgsm values, unless the `_topup` suffix is removed. How do we plan to distinguish between topups and non topups in the same run on the samplesheet if the `_topup` attribute is removed?

Reply:
That's exactly my question!
If the `_topup` is removed, there is no difference as far as I can see. And as we agreed to handle those separately, I don't see why/how we could merge "non-topups" split over lanes.
I can't say that I understand the subtle differences in terms of QC/interpretation, and I don't really mind either way. 
However, I think we should be consistent and right now we are treating different FASTQs with separate QC runs. In principle one could also just use one of those FASTQs when deciding to merge data for analysis, or the run could have an issue with just one lane,... so from a technical point of view, I'd keep them separate. It's easy and flexible.
Happy to discuss with a wider audience though!

Reply:
Does this make the coverage logic a little more complicated though? 

i.e 3 of 4 of the lanes have completed the qc pipeline and are now above our threhsold of 30x coverage. How do we know to wait for the fastq on the last lane to complete processing before continuing with the downstream germline pipeline?

Reply:
That's why we wait until all jobs of the same run have finished.
I.e. the logic to kick off T/N workflows only starts after all QC jobs of the same run have finished. Initially that was to capture cases where tumor and normal are on the same run (we can't make the right decision just after one of them, we have to wait until all are finished before we can say if the T/N workflow can start), but it applies the same here.

Reply:
I have no idea if there is a difference (other than naming) between one library split across multiple lanes or a library and a corresponding topup on the same run. If we remove the `_topup` extension in the future, I don't think we have a way to distinguish them (not sure there is an actual difference).
So, as I see it we have two options: either treat every FASTQ/lane on its own or merge on library ID not matter topup or not.

The former may give us a bit more flexibility and would correspond to FASTQs (pairs). The latter may be more intuitive to users and would mean QC/yield values would already be aggregated (although I am not sure on a level that is expected/most useful).

Perhaps one thing to consider is be the MultiQC report of the BCL Convert workflow. it does merge on library ID, right (which means it would merge the topup when we remove that extension)? But it also shows stats per lane?

Another thing is that the current plan was to distinguish libraries from topups by library_id + run + lane. That means we currently don't have a library + run level record to store results from merged libs.

Reply:
> Perhaps one thing to consider is be the MultiQC report of the BCL Convert workflow. it does merge on library ID, right (which means it would merge the topup when we remove that extension)? But it also shows stats per lane?

It will merge on the `rgsm` value. The output value of `rglb` from bclconvert is `UnknownLibrary`. We are unable to put this in v2 samplesheets. The rgsm and rglb are split on the portal side. In terms of top up coverage, I don't think it matters if we combine or not for multiqc (it would probably be more convenient if it did as we're looking for aggregate coverage anyway).

I think it's worth reconsidering if a topup on the same run is even really a top up or merely a 'same-run-extension' of the current sample.

Reply:
Hm... back to square one then.

I don't think a topup on the same run is a real topup (after all how do you know you need a topup before you have the yield?).

So going forward, we'd merge on library ID (without extension, treating everything on the same run with the same (library) ID as one)?

That would then mean we don't need to keep track of which lane a lib was on, i.e. we'd define uniqueness on library_id + run, and that's where we record qc/yield values.

That would then also mean we couldn't really split libraries apart again (i.e. by lane), right? Not sure that's ever going to be a use case though....

Reply:
Resolved by Zoom quick chat. TL;DR we merge by rglb / pure library ID.